std/experimental/color/rgb.d

```diff
@@ -0,0 +1,852 @@
+
+    // i think this is the linear conversion coefficients
+    wiki = 0.299r + 0.587g + 0.114b
+    ps: (0.3)+(0.59)+(0.11)
+
+    // but gimp source seems to use this one... perhaps this performs an sRGB space estimate, or is THIS the linear one?
+    gimp: 0.21 R + 0.72 G + 0.07 B
+    another = 0.2126 R + 0.7152 G + 0.0722 B   (apparently in linear space?)
+*/
+    return toGamma!colorSpace(T(0.299)*toLinear!colorSpace(r) + T(0.587)*toLinear!colorSpace(g) + T(0.114)*toLinear!colorSpace(b));
+}
+T toGrayscale(RGBColorSpace colorSpace = RGBColorSpace.sRGB, T)(T r, T g, T b) pure if(isIntegral!T)
+{
+    alias F = FloatTypeFor!T;
+    return convertPixelType!T(toGrayscale!colorSpace(convertPixelType!F(r), convertPixelType!F(g), convertPixelType!F(b)));
+}
+
+
+// helpers to parse color components from color component string
+template CanFind(string s, char c)
+{
+    static if(s.length == 0)
+        enum CanFind = false;
+    else
+        enum CanFind = s[0] == c || CanFind!(s[1..$], c);
+}
+template AllIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum AllIn = true;
+    else
+        enum AllIn = CanFind!(s, chars[0]) && AllIn!(s, chars[1..$]);
+}
+template AnyIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum AnyIn = false;
+    else
+        enum AnyIn = CanFind!(s, chars[0]) || AnyIn!(s, chars[1..$]);
+}
+template NoneIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum NoneIn = true;
+    else
+        enum NoneIn = !CanFind!(s, chars[0]) && NoneIn!(s, chars[1..$]);
+}
+template NotIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum NotIn = char(0);
+    else static if(!CanFind!(s, chars[0]))
+        enum NotIn = chars[0];
+    else
+        enum NotIn = NotIn!(s, chars[1..$]);
+}
+
+unittest
+{
+    static assert(CanFind!("string", 'i'));
+    static assert(!CanFind!("string", 'x'));
+    static assert(AllIn!("string", "sgi"));
+    static assert(!AllIn!("string", "sgix"));
+    static assert(AnyIn!("string", "sx"));
+    static assert(!AnyIn!("string", "x"));
+    static assert(NoneIn!("string", "xyz"));
+    static assert(!NoneIn!("string", "sxyz"));
+}
+
+
+// shorthand alias for all components
+alias AllComponents = TypeTuple!('l','r','g','b','a');
+
+// build mixin code to perform expresions per-element
+template ComponentExpression(string expression, char component, string op)
+{
+    template BuildExpression(string e, char c, string op)
+    {
+        static if(e.length == 0)
+            enum BuildExpression = "";
+        else static if(e[0] == '_')
+            enum BuildExpression = c ~ BuildExpression!(e[1..$], c, op);
+        else static if(e[0] == '#')
+            enum BuildExpression = op ~ BuildExpression!(e[1..$], c, op);
+        else
+            enum BuildExpression = e[0] ~ BuildExpression!(e[1..$], c, op);
+    }
+    enum ComponentExpression =
```

Comment:
> >I arrived at this in combat of unruly spaghetti. I'm open to suggestions...
>
> Composition, composition, composition :)

How would that look in this case? Don't suggest I create more wrapper types to do a mul or an add...
As much as Andrei loves composition, I take a very conservative approach to employment of composition. Unless it offers tangible benefit, I feel the loss of clarity and debug-ability to fail the balance in my mind.

> > unittests should help here.
>
> Ironically, this is one situation where unit tests will be less helpful, since the problem is that > incorrect code will be silently accepted in unexpected situations, which are hard to foresee.

Can you suggest an example?
What would I change it to?